pallets/sto/src/benchmarking.rs

```diff
@@ -0,0 +1,215 @@
+
+        let offering_ticker = Ticker::try_from(&[b'A'][..]).unwrap();
+        let raise_ticker = Ticker::try_from(&[b'B'][..]).unwrap();
+        create_asset::<T>(alice.origin(), offering_ticker, 1_000_000)?;
+        create_asset::<T>(alice.origin(), raise_ticker, 1_000_000)?;
+
+        add_compliance::<T>(alice.origin(), offering_ticker)?;
+        add_compliance::<T>(alice.origin(), raise_ticker)?;
+
+        let venue_id = <Settlement<T>>::venue_counter();
+        <Settlement<T>>::create_venue(
+            alice.origin().into(),
+            VenueDetails::default(),
+            vec![alice.account()],
+            VenueType::Sto
+        )?;
+    }: _(
+            alice.origin(),
+            alice_portfolio,
+            offering_ticker,
+            alice_portfolio,
+            raise_ticker,
+            generate_tiers::<T>(i),
+            venue_id,
+            None,
+            None
+        )
+    verify {
+        ensure!(FundraiserCount::get(offering_ticker) > 0, "create_fundraiser");
+    }
+
+    invest {
+        // Rule complexity
+        let c in 1 .. T::MaxConditionComplexity::get() as u32;
+
+        let alice = <UserBuilder<T>>::default().generate_did().build("alice");
+        let alice_portfolio = PortfolioId::default_portfolio(alice.did());
+        let bob = <UserBuilder<T>>::default().generate_did().build("bob");
+        let bob_portfolio = PortfolioId::default_portfolio(bob.did());
+
+        let offering_ticker = Ticker::try_from(&[b'A'][..]).unwrap();
+        let raise_ticker = Ticker::try_from(&[b'B'][..]).unwrap();
+        create_asset::<T>(alice.origin(), offering_ticker, 1_000_000)?;
+        create_asset::<T>(alice.origin(), raise_ticker, 1_000_000)?;
+
+        let t_issuer = UserBuilder::<T>::default().generate_did().build("TrustedClaimIssuer");
+        let trusted_issuer = TrustedIssuer::from(t_issuer.did());
+
+        compliance_setup::<T>(c, offering_ticker, alice.origin(), alice.did(), bob.did(), trusted_issuer.clone());
+        compliance_setup::<T>(c, raise_ticker, alice.origin(), bob.did(), alice.did(), trusted_issuer);
+
+        let venue_id = <Settlement<T>>::venue_counter();
+        <Settlement<T>>::create_venue(
+            alice.origin().into(),
+            VenueDetails::default(),
+            vec![alice.account()],
+            VenueType::Sto
+        )?;
+
+        <Asset<T>>::unsafe_transfer(
+            alice_portfolio,
+            bob_portfolio,
+            &raise_ticker,
+            1_000_000.into()
+        )?;
+
+        <Sto<T>>::create_fundraiser(
+            alice.origin().into(),
+            alice_portfolio,
+            offering_ticker,
+            alice_portfolio,
+            raise_ticker,
+            generate_tiers::<T>(MAX_TIERS as u32),
+            venue_id,
+            None,
+            None
+        )?;
+    }: _(
+            bob.origin(),
+            bob_portfolio,
+            bob_portfolio,
+            offering_ticker,
+            0,
```

Comment:
Won't this trigger branch-taken in `if remaining == 0.into()  { break }` therefore ensuring the best case complexity?

Reply:
That's the fundraiser id not the amount invested

Reply:
Would be good to newtype this (let's make a separate story) to make that easier to read?

Reply:
Story: https://polymath.atlassian.net/browse/MESH-1479